libr/debug/trace.c

```diff
@@ -98,16 +99,135 @@ R_API RDebugTracepoint *r_debug_trace_get (RDebug *dbg, ut64 addr) {
+}
+
+// Copy from visual to avoid circular dependency
+void r_core_visual_list(RCore *core, RList *list, ut64 seek, ut64 len, int width, int use_color) {
+	ut64 mul, min = -1, max = -1;
+	RListIter *iter;
+	ListInfo *info;
+	int j, i;
+	RIO *io = core->io;
+	width -= 80;
+	if (width < 1) {
+		width = 30;
+	}
+
+	r_list_foreach (list, iter, info) {
+		if (min == -1 || info->pitv.addr < min) {
+			min = info->pitv.addr;
+		}
+		if (max == -1 || info->pitv.addr + info->pitv.size > max) {
+			max = info->pitv.addr + info->pitv.size;
+		}
+	}
+	mul = (max - min) / width;
+	if (min != -1 && mul > 0) {
+		const char * color = "", *color_end = "";
+		i = 0;
+		r_list_foreach (list, iter, info) {
+			if (use_color && info->perm != -1) {
+				color_end = Color_RESET;
+				if ((info->perm & R_PERM_X) && (info->perm & R_PERM_W)) { // exec & write bits
+					color = r_cons_singleton ()->context->pal.graph_trufae;
+				} else if ((info->perm & R_PERM_X)) { // exec bit
+					color = r_cons_singleton ()->context->pal.graph_true;
+				} else if ((info->perm & R_PERM_W)) { // write bit
+					color = r_cons_singleton ()->context->pal.graph_false;
+				} else {
+					color = "";
+					color_end = "";
+				}
+			} else {
+				color = "";
+				color_end = "";
+			}
+			if (io->va) {
+				io->cb_printf ("%05d%c %s0x%08"PFMT64x"%s |", i,
+						r_itv_contain (info->vitv, seek) ? '*' : ' ',
+						color, info->vitv.addr, color_end);
+			} else {
+				io->cb_printf ("%05d%c %s0x%08"PFMT64x"%s |", i,
+						r_itv_contain (info->pitv, seek) ? '*' : ' ',
+						color, info->pitv.addr, color_end);
+			}
+			for (j = 0; j < width; j++) {
+				ut64 pos = min + j * mul;
+				ut64 npos = min + (j + 1) * mul;
+				if (info->pitv.addr < npos && (info->pitv.addr + info->pitv.size) > pos) {
+					io->cb_printf ("#");
+				} else {
+					io->cb_printf ("-");
+				}
+			}
+			if (io->va) {
+				io->cb_printf ("| %s0x%08"PFMT64x"%s %s %6s %s\n",
+					color, r_itv_end (info->vitv), color_end,
+					(info->perm != -1)? r_str_rwx_i (info->perm) : "   ",
+					(info->extra)?info->extra : "    ",
+					(info->name)?info->name : " ");
+			} else {
+				io->cb_printf ("| %s0x%08"PFMT64x"%s %s %6s %s\n",
+					color, r_itv_end (info->pitv), color_end,
+					(info->perm != -1)? r_str_rwx_i (info->perm) : "   ",
+					(info->extra)?info->extra : "      ",
+					(info->name)?info->name : "");
+			}
+			i++;
+		}
+		/* current seek */
+		if (i > 0 && len != 0) {
+			if (seek == UT64_MAX) {
+				seek = 0;
+			}
+			io->cb_printf ("=>     0x%08"PFMT64x" |", seek);
+			for (j = 0; j < width; j++) {
+				io->cb_printf (
+					((j * mul) + min >= seek &&
+					 (j * mul) + min <= seek+len)
+					?"^" : "-");
+			}
+			io->cb_printf ("| 0x%08"PFMT64x"\n", seek+len);
+		}
+	}
+}
+
 R_API void r_debug_trace_list (RDebug *dbg, int mode) {
 	int tag = dbg->trace->tag;
 	RListIter *iter;
+	bool flag = false;
+	RList *info_list = r_list_new ();
+	if (!info_list && mode == '=') {
```

Comment:
thats probably never going to happen. info_list. better use || instead of &&

Reply:
The idea is that info_list is only required when the mode is dt=. So I think && is the right op here?